### PR TITLE
try to keep ids as low as possible

### PR DESCRIPTION
**Pull request by condret**
this should halp us saving memory, when we open bigger amount of fds at once

### CODE REVIEW

File: libr/util/idpool.c [hunk 1]
```diff
@@ -31,10 +48,12 @@ R_API bool r_id_pool_grab_id(RIDPool* pool, ut32* grabber) {
...
 	if (pool->freed_ids) {
-		ut32 grab = (ut32) (size_t)r_queue_dequeue (pool->freed_ids);
-		*grabber = (ut32) grab;
-		if (r_queue_is_empty (pool->freed_ids)) {
-			r_queue_free (pool->freed_ids);
+		RBTreeIter iter = r_rbtree_first (pool->freed_ids);
+		*grabber = (ut32)(size_t)(iter.path[iter.len-1]->data);
+//		eprintf ("deleting %d from the tree\n", *grabber);
+		r_rbtree_delete (pool->freed_ids, (void *)(size_t)(*grabber), NULL);
+		if (r_rbtree_size (pool->freed_ids) == 1) {
```
Comment (MaskRay) [inline libr/util/idpool.c]: I think there is no need to free the tree. This simplifies the logic here and in the _delete function.

File: libr/util/idpool.c [hunk 2]
```diff
@@ -53,18 +72,34 @@ R_API bool r_id_pool_kick_id(RIDPool* pool, ut32 kick) {
 	}
 	if (kick == (pool->next_id - 1)) {
 		pool->next_id--;
+		if (pool->freed_ids) {
+			RBTreeIter iter = r_rbtree_last (pool->freed_ids);
+			while ((iter.len) && ((pool->next_id - 1) <= (ut32)(size_t)(iter.path[iter.len-1]->data))) {
```
Comment (MaskRay) [inline libr/util/idpool.c]: I should create an API to make fetching the data associated with the current position easier `iter.path[iter.len-1]->data`

I have an idea to make rbtree intrusive ...

struct YourStruct { struct r_rbtree_t rb; ... }

File: libr/util/idpool.c [hunk 3]
```diff
@@ -53,18 +72,34 @@ R_API bool r_id_pool_kick_id(RIDPool* pool, ut32 kick) {
...
+//				eprintf ("element in tree %d\n", (ut32)(size_t)(iter.path[iter.len-1]->data));
+				pool->next_id--;
+				r_rbtree_iter_prev (&iter);
+			}
+			pool->freed_ids->cmp = (RBTreeComparator)_del_cmp;
```
Comment (MaskRay) [inline libr/util/idpool.c]: I think there is no need to change the comparison function.

File: libr/util/idpool.c [hunk 4]
```diff
@@ -53,18 +72,34 @@ R_API bool r_id_pool_kick_id(RIDPool* pool, ut32 kick) {
...
+			while (r_rbtree_delete (pool->freed_ids, (void *)(size_t)(pool->next_id - 1), NULL)) {}
```
Comment (MaskRay) [inline libr/util/idpool.c]: I guess this line never deletes an element.

**Comment by MaskRay** [comment]
> I guess this line never deletes an element.

Sorry, this is incorrect.

I think the logic can be simplified to:

```c
while (next_id > 0 && next_id-1 in tree) next_id--;
while (tree_not_empty && tree_last_key >= next_id) delete it;
```

File: libr/util/idpool.c (hunk 3 above)
Comment (condret) [reply libr/util/idpool.c]: well first i find out what the new next_id should be, then I delete all the ids that are stored in the tree, that are bigger or equal to that. I don't do the delition, in the same place, while figuring out the new next_id, bc I fear to break the iterator, when deleting in iterations

**Comment by condret** [comment]
since the old tree api is gone

**Pull request closed without merging**